src/iosource/Manager.cc

```diff
@@ -157,25 +157,24 @@ void Manager::FindReadySources(ReadySources* ready)
 		if ( iosource->IsOpen() )
 			{
 			double next = iosource->GetNextTimeout();
-			bool added = false;
 
 			if ( timeout == -1 || (next >= 0.0 && next < timeout) )
 				{
 				timeout = next;
 				timeout_src = iosource;
-
-				// If a source has a zero timeout then it's ready. Just add it to the
-				// list already. Only do this if it's not time to poll though, since
-				// we don't want things in the vector passed into Poll() or it'll end
-				// up inserting duplicates.
-				if ( timeout == 0 && ! time_to_poll )
-					{
-					added = true;
-					ready->push_back({timeout_src, -1, 0});
-					}
 				}
 
-			if ( iosource == pkt_src && ! added )
+			// If a source has a zero timeout then it's ready. Just add it to the
+			// list already. Only do this if it's not time to poll though, since
+			// we don't want things in the vector passed into Poll() or it'll end
+			// up inserting duplicates. A source with a zero timeout that was not
+			// selected as the timeout_src can be safely added, whether it's time
+			// to poll or not though.
+			if ( next == 0 && (! time_to_poll || iosource != timeout_src) )
```

Comment:
I'll put the Poll() fix for timeout_src and timeout == 0 into #2866. I think I have an idea.

> I like the idea of having the possibility to combine both. 

After staring, it should actually be possible to do so.

Reply:
I might be missing something but I don't see this reflected in #2866 as well. Imho, at least documenting the implicit assumption would be curial.

Reply:
> An IO source can actually be using timeouts and be FD based. It may trigger double processing currently, though that seem least problematic.

Maybe as a bit of a follow-up, because I was rather brief. To me IO sources mixing FD and timeouts seem mostly uncharted territory. My latest read was that they should work out mostly now, though `Process()` possibly called twice when the FD is ready, they returned a timeout of 0.0, it being time_to_poll and there being another IO source with a timeout of 0.0 that was actually selected as the timeout_src for Poll(). And I'd leave it open that we prevent that in the future so they are processed just once.

IMO, once we find limitations and things that don't work quite as needed or expected, should improve from there and if things appear consistent and reasonable, have tests and document it (the above isn't quite there yet I think).

There's also some behaviors that I find rather surprising or questionable. Most prominently, reading a pcap without `suspend_processing()` that has less than 100 packets, FD based IO sources won't ever get polled or processed within the main-loop. With more packets available, FD based IO source processing happens only every 100 packets and in live operating mode 10 packets - which may or may not be the right behavior in certain use-cases.